subsys/net/openthread/rpc/server/ot_rpc_udp_server.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <nrf_rpc/nrf_rpc_serialize.h>
+#include <ot_rpc_ids.h>
+#include <ot_rpc_types.h>
+#include <ot_rpc_common.h>
+
+#include <nrf_rpc_cbor.h>
+
+#include <openthread/error.h>
+#include <openthread/instance.h>
+
+#include <openthread/udp.h>
+
+#include <zephyr/net/openthread.h>
+#include <zcbor_decode.h>
+#include <string.h>
+
+#include "ot_rpc_message_server.h"
+
+#define OT_UDP_SOCKETS 8
+
+typedef struct nrfUdpSocket {
+	ot_socket_key mKey;
+	otUdpSocket mSocket;
+} nrfUdpSocket;
+
+static nrfUdpSocket sSockets[OT_UDP_SOCKETS];
+
+static nrfUdpSocket *nrf_udp_find_socket(ot_socket_key key)
+{
+	for (size_t i = 0; i < OT_UDP_SOCKETS; i++) {
+		if (key == sSockets[i].mKey) {
+			return &sSockets[i];
+		}
+	}
+
+	return NULL;
+}
+
+static nrfUdpSocket *nrf_udp_alloc_socket(ot_socket_key key)
+{
+	nrfUdpSocket *ret = nrf_udp_find_socket(key);
+
+	if (ret != NULL) {
+		return ret;
+	}
+
+	ret = nrf_udp_find_socket(0);
+
+	if (ret != NULL) {
+		memset(ret, 0, sizeof(nrfUdpSocket));
+		ret->mKey = key;
+	}
+
+	return ret;
+}
+
+static void nrf_udp_free_socket(ot_socket_key key)
+{
+	nrfUdpSocket *socket = nrf_udp_find_socket(key);
+
+	if (socket != NULL) {
+		memset(socket, 0, sizeof(nrfUdpSocket));
+	}
+}
+
+static void handle_udp_receive(void *aContext, otMessage *aMessage,
+			       const otMessageInfo *aMessageInfo)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	nrfUdpSocket *socket = (nrfUdpSocket *)aContext;
+	ot_msg_key msg_key = ot_reg_msg_alloc();
+	ot_socket_key soc_key = socket->mKey;
+
+	ot_msg_set(msg_key, aMessage);
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx,
+			   sizeof(soc_key) + sizeof(msg_key) + sizeof(otMessageInfo) + 16);
+
+	if (!zcbor_uint_encode(ctx.zs, &soc_key, sizeof(soc_key))) {
+		NRF_RPC_CBOR_DISCARD(&ot_group, ctx);
+		goto exit;
+	}
+
+	if (!zcbor_uint_encode(ctx.zs, &msg_key, sizeof(msg_key))) {
+		NRF_RPC_CBOR_DISCARD(&ot_group, ctx);
+		goto exit;
+	}
+
+	if (!ot_rpc_encode_message_info(&ctx, aMessageInfo)) {
+		NRF_RPC_CBOR_DISCARD(&ot_group, ctx);
+		goto exit;
+	}
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, OT_RPC_CMD_UDP_RECEIVE_CB, &ctx);
+
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+
+	ot_msg_free(msg_key);
```

Comment:
In case of encoding error, there will be memory leakage.

Reply:
fixed.